community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/PhysicalLogFile.java

```diff
@@ -149,7 +149,13 @@ private PhysicalLogVersionedStoreChannel rotate( LogVersionedStoreChannel curren
          * current log file and replay everything. That's unnecessary but totally ok.
          */
         long newLogVersion = logVersionRepository.incrementAndGetVersion();
-        currentLog.flush();
+        /*
+         * Rotation can happen at any point, although not concurrently with an append,
+         * although an append may have (most likely actually) left at least some bytes left
+         * in the buffer for future flushing. Flushing that buffer now makes the last appended
+         * transaction complete in the log we're rotating away. Awesome.
+         */
+        writer.emptyBufferIntoChannelAndClearIt().flush();
```

Comment:
Hmm, actually this turns out to be two separate things. Flushing can only be done by calling `emptyBufferIntoChannelAndClearIt()` unless you have a direct reference to the underlying channel, which is what rotation code in PhysicalLogFile did wrong.